Remove MavenFilteringHack

jakelandis opened the pull request:
@mark-vieira / @breskeby  - Can you take a quick peek at the direction here ?  I am in need of the ability to apply the functionality of MavenFilteringHack when copying YAML tests around, but the CopyRestTestsTask isn't a proper Copy task. To avoid duplication and re-use of the Maven standard `${`foo`}`, this change will also remove the MavenFilteringHack and update all placeholder tokens to `@`foo`@` which is the Ant default value which allows for less custom code. 

Note - This is a draft PR until I double check that I got all of the `${`foo`}` (I did this draft pretty quickly) and do a bit more manual testing. 

Note 2 - for my needs I could get away with just the updates to the CopyRestTestTask ... but that would require either 2 ways to represent placeholders or duplicating/re-using the Maven conventions. 

Closes https://github.com/elastic/elasticsearch/issues/71434

jakelandis committed:
Remove MavenFilteringHack

mark-vieira commented on distribution/src/config/jvm.options:
```diff
@@ -72,10 +72,10 @@
+@error.file@
 
 ## GC logging
 -Xlog:gc*,gc+age=trace,safepoint:file=${loggc}:utctime,pid,tags:filecount=32,filesize=64m
```
Missed one. Need to update `${loggc}` here.

joegallo replied:
f3fb120ef100c343f02069a64b2d3a7e27f346b1

mark-vieira commented on distribution/archives/integ-test-zip/build.gradle:
```diff
@@ -17,7 +19,7 @@ group = "org.elasticsearch.distribution.integ-test-zip"
 tasks.named("processTestResources").configure {
   inputs.properties(project(':distribution').restTestExpansions)
-  MavenFilteringHack.filter(it, project(':distribution').restTestExpansions)
+  it.filter("tokens" : project(':distribution').restTestExpansions, ReplaceTokens.class)
```
FWIW, we can just do `filter` here instead of `it.filter`. Can we update all the other build scripts appropriately? Unnecessarily qualifying this is just kind of confusing.

joegallo replied:
a2efdd9

mark-vieira commented on distribution/archives/integ-test-zip/build.gradle:
```diff
@@ -17,7 +19,7 @@ group = "org.elasticsearch.distribution.integ-test-zip"
 
 tasks.named("processTestResources").configure {
   inputs.properties(project(':distribution').restTestExpansions)
```
I believe we can remove all these explicit input declarations as well now. Since we are passing the expansions map to `filter()` the copy spec should be tracking those items as an input for us. Might want to confirm though by changing a map value and confirming the task is no longer up-to-date.

joegallo replied:
That doesn't play out in my testing of this. If I remove those lines, then the tasks are incorrectly considered `up-to-date` even when the expansions map changes. Or, at least, that's how it's looking in my runs.

mark-vieira replied:
:shrug: fair enough, there's no real disadvantage to be explicit here so I'm good with this as-is.

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/CopyRestTestsTask.java:
```diff
@@ -46,6 +49,7 @@
     private static final String REST_TEST_PREFIX = "rest-api-spec/test";
     private final ListProperty<String> includeCore;
     private final ListProperty<String> includeXpack;
+    private Map<String, String> substitutions;
```
Should we add test coverage for this?

joegallo replied:
Sure thing, 4797e9c

joegallo committed:
Merge branch 'master' into remove_MavenFilterHack

joegallo committed:
Fix some missing substitutions in distribution/

This is enough to get some of the yamlRestTests passing for me.

joegallo committed:
Whitespace

joegallo committed:
s/it.filter/filter/g

joegallo committed:
These need to be strings

joegallo committed:
Revert these changes

They're runtime substitutions within log4j2 itself rather than build
time substitutions.

joegallo committed:
Fix repository-gcs tests

joegallo committed:
Fix repository-s3 tests

joegallo committed:
Fix some of the distribution scripts

breskeby reviewed (approved):
hey Jake this looks good for me. Nothing to add except what mark said. some basic test coverage would be good to have IMO

joegallo committed:
Organize imports

joegallo committed:
Whitespace

joegallo committed:
Add substitutions to the rest resources integTest

joegallo committed:
Merge branch 'master' into remove_MavenFilterHack

elasticmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

jakelandis commented on distribution/build.gradle:
```diff
@@ -215,7 +215,7 @@ void copyLog4jProperties(TaskProvider buildTask, Project module) {
 
 ext.restTestExpansions = [
-  'expected.modules.count': 0,
+  'expected.modules.count': "0", //TODO: is it safe to make this a string ?
```
This todo might be worth a quick follow u on it's usage to ensure we don't break something with distribution.

joegallo replied:
19762bb

jakelandis commented on plugins/repository-s3/build.gradle:
```diff
@@ -153,10 +153,10 @@ tasks.named("processYamlRestTestResources").configure {
     'ecs_bucket'              : s3ECSBucket,
     'ecs_base_path'           : s3ECSBasePath,
-    'disable_chunked_encoding': s3DisableChunkedEncoding,
+    'disable_chunked_encoding': s3DisableChunkedEncoding.toString(), //TODO: ensure string is OK here
```
ditto as above on the TODO

joegallo replied:
19762bb

jakelandis commented on qa/smoke-test-plugins/build.gradle:
```diff
@@ -31,11 +31,11 @@ testClusters.matching { it.name == "integTest" }.configureEach {
 
 ext.expansions = [
-  'expected.plugins.count': pluginPaths.size()
+  'expected.plugins.count': pluginPaths.size().toString()
```
here too.

joegallo replied:
19762bb

jakelandis commented on x-pack/qa/smoke-test-plugins-ssl/build.gradle:
```diff
@@ -85,18 +85,14 @@ testClusters.matching { it.name == "integTest" }.configureEach {
-
-ext.expansions = [
-  'expected.plugins.count': pluginPaths.size()
+  'expected.plugins.count': pluginPaths.size().toString()
```
^

joegallo replied:
19762bb

jakelandis commented on x-pack/qa/smoke-test-plugins/build.gradle:
```diff
@@ -21,12 +21,12 @@ def pluginPaths = project(':plugins').getChildProjects().findAll { pluginName, p
 
 ext.expansions = [
-        'expected.plugins.count': pluginPaths.size()
+        'expected.plugins.count': pluginPaths.size().toString()
```
^

joegallo replied:
19762bb

joegallo committed:
Convert to Map<String, String> at invocation time

For the groovy side of things, we'd rather have Map<String, Object>,
but ReplaceTokens itself is quite picky and requires Map<String,
String> (well, it requires at runtime that the values in the map are
all Strings). Anyway, this allows us to be loose with the types in our
build files, but still have the required type for ReplaceTokens.

It's a compromise.

joegallo committed:
Simplify this test just a little

joegallo committed:
Merge branch 'master' into remove_MavenFilterHack

joegallo committed:
These should always be strings already

So we don't need to the toString, but it doesn't hurt to throw an
exception if somebody violates our expectations.

Also the `expansions =` assignment seems unnecessary to me, so I
dropped it.

joegallo committed:
Tweak and add a comment

joegallo committed:
Remove stopping.timeout

It's unused now as of #51716.

joegallo committed:
Drop the type check and go back to coercing to String

But add a comment, too

jakelandis commented:
LGTM
